docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,154 @@
+**<h1>Train API Proposal for LLMs</h1>**
+
+**<h3>Authors:</h3>**
+
+* Deepanker Gupta (**[@deepanker13](https://github.com/deepanker13)**), Nutanix
+* Johnu George (**[@johnugeorge](https://github.com/johnugeorge)**), Nutanix
+
+**<h3>Status</h3>**
+
+* 10 Nov 2023 (v1)
+
+**<h3>Abstract</h3>**
+
+LLMs are being widely used for generative AI tasks and as their adoption is increasing across various domains, the need for LLMOps in the Kubernetes environment has also risen. There is a need to fine-tune these large models on task-specific datasets. The user/data scientist should be able to do this simplistically in the Kubernetes environment using the Kubeflow training operator SDK without any infrastructure knowledge.
+
+**<h3>Background</h3>**
+
+Currently, there are two flows for data scientists to start using Training operator for their distributed training needs. 
+
+**<h4>Traditional method</h4>**
+
+1. The end-user has to write a custom script that includes data fetching, model loading, and a training loop using pytorch libraries.
+2. Create an image and push it to a registry.
+3. Using the Kubeflow train sdk to create worker spec and master spec which uses the above image and using these two specs create a pytorch job.
+
+This method requires infrastructure knowledge including image creation, registry access and training job spec. This makes it difficult for data scientists to purely focus at the application layer.
+
+**<h4>Higher level SDK</h4>**
+
+To provide a better user experience, a new higher level SDK was added in[ https://github.com/kubeflow/training-operator/pull/1659](https://github.com/kubeflow/training-operator/pull/1659) which avoided the need of image creation. Instead of creating the image, we can directly pass the custom script to train_func parameter and final pytorch job is created
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
```

Comment:
I think create_job can be used by a user who is interested in having finer control over the training job, where as train is a very high level api

Reply:
* I think that `create_job` is more general than `train`, while this repo this under the scope `training` already.

* If we take `create_job` for more control flow, is that mean we want to do things like `create_job` to initialize a job and then `start_job` to start it etc.

Reply:
But would it be simpler to just add more arguments to the `train` API that `create_job` API supports today ?
E.g. in Katib we have `tune` API where user can use it for finer control and for high level API: https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L135-L156

Also, as @kuizhiqing mentioned Training Operator currently don't support ability to Create Job and then Start Job.

Reply:
@andreyvelich we can take this up in the phase 2 of the implementation. When we have first version of the train API, we can think about merging parameters from `create_job` API .    I feel, it is better not to over complicate at this point and keep API as simple as possible for the data scientists.  We will see the adoption and add features accordingly.

Reply:
@andreyvelich I fully agree with providing only `train` API. 
However, I think that removing `create_job` and introducing `train` API is out of scope of this proposal.
Ideally, we should improve our SDK before working on this feature.

Reply:
Ah, I misunderstood this proposal. This proposal tries to introduce the `train` API.

Reply:
I think we can take 2 ways like this:

1. Implement `train` API
2. Expand `create_job` API

If we can take option 1, as @andreyvelich said, it would be better to consolidate `create_job` API to `train` API.
Maybe, we should work on consolidating before we work on the LLM feature.

If we can take option 2, we can just add some arguments to the existing `create_job` API.

It means I think we shouldn't provide libraries similar to existing libraries.

Reply:
@johnugeorge What do you think about implementation cost ? How much effort it would be to create `train` API first, and after that merge all parameters from `create_job` API to `train` API ?

Reply:
@andreyvelich create_job is a low level flexible API meant for all kinds of Training Job Types. eg: It handles arbitrary python functions, supports different Kubeflow training job types, supports framework specific concepts like PS, chief etc.  

However, we are tracking a provider specific higher level API here. We needn't bring framework specific concepts to this API. Internal implementation needs to manage how the higher level API is mapped into framework specific code. See  [MPIJob discussion](https://github.com/kubeflow/training-operator/pull/1945#issuecomment-1824211828).   Also, I am not sure if it is a good idea to bring Kubernetes specific Job object types here. This will defeat the purpose of getting a data scientist specific API.

Reply:
Makes sense @johnugeorge.
For the initial implementation let's have this 2 APIs: `train` for fine-tuning and `create_job` for any arbitrary training that users want to run.
After the implementation, we can discuss whether we should expose only single API for Training Operator users if they want to train their models.

Reply:
SGTM.

Reply:
SGTM

Reply:
@johnugeorge  @andreyvelich @tenzen-y  I have raised a pr https://github.com/kubeflow/training-operator/pull/1954, Please have a look at the train function and share your feedback. No changes will be required in create_job function.